### PR TITLE
SDC Expansion Resource

### PR DESCRIPTION
**Pull request by Krunal-Thakkar**
SDC Expansion Resource

That will handle the Parse CSV, Validate MDM, and Installation Functions.
image

### CODE REVIEW

File: docs/resources/sdc_expansion.md [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+```terraform
+# Command to run this tf file : terraform init && terraform plan && terraform apply
+# Create, Update, Delete is supported for this resource
+# To import , check volume_resource_import.tf for more info
+# To create / update, either storage_pool_id or storage_pool_name must be provided
+# Also , to create / update, either protection_domain_id or protection_domain_name must be provided
+# name, size is the required parameter to create or update
+# other  atrributes like : capacity_unit, volume_type, use_rm_cache, compression_method, access_mode, remove_mode, sdc_list are optional 
+# To check which attributes of the snapshot can be updated, please refer Product Guide in the documentation
+
```
Comment (AnikaAgiwal2711) [inline docs/resources/sdc_expansion.md:41]: All the above comments is for Volume. Please update for sdc_expansion resource

File: docs/resources/sdc_expansion.md [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+resource "powerflex_sdc_expansion" "test-csv2" {
+	mdm_ip = "10.247.103.160"
+	mdm_password = "ABCD"
+	lia_password="ABCD"
+	csv_detail = [
+		{
+			ip = "10.247.103.160"
```
Comment (AnikaAgiwal2711) [inline docs/resources/sdc_expansion.md:48]: Please remove the ip, password or any sensitive information . Applicable for all such instances

File: docs/resources/sdc_expansion.md (hunk 1 above)
Comment (Krunal-Thakkar) [reply docs/resources/sdc_expansion.md:41]: Updated.

File: docs/resources/sdc_expansion.md (hunk 2 above)
Comment (Krunal-Thakkar) [reply docs/resources/sdc_expansion.md:48]: Updated

File: powerflex/sdc_expansion_resource_schema.go [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+	Description:         "This resource can be used to add the SDC in PowerFlex Cluster.",
+	MarkdownDescription: "This resource can be used to add the SDC in PowerFlex Cluster.",
+	Attributes: map[string]schema.Attribute{
+		"csv_detail": csvSchema,
+		"mdm_ip": schema.StringAttribute{
+			Description:         "MDM Server IPs. User can provide Primary and Secondary MDM IP comma seperated",
+			MarkdownDescription: "MDM Server IPs. User can provide Primary and Secondary MDM IP comma seperated",
+			Required:            true,
+		},
+		"mdm_password": schema.StringAttribute{
```
Comment (shenda1) [inline powerflex/sdc_expansion_resource_schema.go:52]: Is MDM password will be the same for all MDMs?

File: powerflex/sdc_expansion_resource_schema.go [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+			Description:         "The ID of the package.",
+			Computed:            true,
+			MarkdownDescription: "The ID of the package.",
+		},
+	},
+}
+
+// csvSchema - varible holds schema for CSV Param Details
+var csvSchema schema.SetNestedAttribute = schema.SetNestedAttribute{
+	Description:         "List OF SDC Expansion Server Details.",
```
Comment (shenda1) [inline powerflex/sdc_expansion_resource_schema.go:77]: List of

File: powerflex/sdc_expansion_resource_schema.go [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+				PlanModifiers: []planmodifier.String{
+					stringDefault(" "),
+				},
+			},
+			"is_sdc": schema.StringAttribute{
+				Description:         "whether this node is SDC or not",
+				Required:            true,
+				MarkdownDescription: "whether this node is SDC or not",
+			},
+			"performance_profile": schema.StringAttribute{
```
Comment (shenda1) [inline powerflex/sdc_expansion_resource_schema.go:110]: I think Performance Profile of SDC would be sufficient?

File: powerflex/sdc_expansion_resource_schema.go (hunk 3 above)
Comment (Krunal-Thakkar) [reply powerflex/sdc_expansion_resource_schema.go:52]: yes, that will be same.

File: powerflex/sdc_expansion_resource_schema.go (hunk 5 above)
Comment (Krunal-Thakkar) [reply powerflex/sdc_expansion_resource_schema.go:112]: Updated.

File: powerflex/sdc_expansion_resource_schema.go (hunk 4 above)
Comment (Krunal-Thakkar) [reply powerflex/sdc_expansion_resource_schema.go]: Updated.

File: powerflex/sdc_expansion_resource_schema.go [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+		},
+		"id": schema.StringAttribute{
+			Description:         "The ID of the package.",
+			Computed:            true,
+			MarkdownDescription: "The ID of the package.",
+		},
+	},
+}
+
+// csvSchema - varible holds schema for CSV Param Details
```
Comment (shenda1) [inline powerflex/sdc_expansion_resource_schema.go:77]: variable

File: powerflex/sdc_expansion_resource_schema.go [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+	PerformanceProfile string
+	SDCName            string
+}
+
+// SDCExpansionResourceSchema - varible holds schema for SDC Expansion
+var SDCExpansionResourceSchema schema.Schema = schema.Schema{
+	Description:         "This resource can be used to add the SDC in PowerFlex Cluster.",
+	MarkdownDescription: "This resource can be used to add the SDC in PowerFlex Cluster.",
+	Attributes: map[string]schema.Attribute{
+		"csv_detail": csvSchema,
```
Comment (shenda1) [inline powerflex/sdc_expansion_resource_schema.go:48]: Can we have some meaningful name like sdc_details?

File: powerflex/sdc_expansion_resource_schema.go [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+	IP                 string
+	Password           string
+	OperatingSystem    string
+	IsMdmOrTb          string
+	IsSdc              string
+	PerformanceProfile string
+	SDCName            string
+}
+
+// SDCExpansionResourceSchema - varible holds schema for SDC Expansion
```
Comment (shenda1) [inline powerflex/sdc_expansion_resource_schema.go:43]: variable. Applicable for all instances.

File: powerflex/sdc_expansion_resource_test.go [hunk 9]
```diff
@@ -0,0 +1,323 @@
...
+			is_mdm_or_tb = "Standby"
+			is_sdc = "Yes"
+   		},
+	]
+}
+`
+
+var ParseCSVConfigUpdate = `
+resource "powerflex_sdc_expansion" "test" {
+	mdm_ip = "10.247.103.160"
```
Comment (shenda1) [inline powerflex/sdc_expansion_resource_test.go:72]: As discussed, use env variables.

Comment (Krunal-Thakkar) [reply powerflex/sdc_expansion_resource_test.go:72]: Updated.

File: powerflex/sdc_expansion_resource_schema.go (hunk 8 above)
Comment (Krunal-Thakkar) [reply powerflex/sdc_expansion_resource_schema.go:43]: Updated.

File: powerflex/sdc_expansion_resource_schema.go (hunk 6 above)
Comment (Krunal-Thakkar) [reply powerflex/sdc_expansion_resource_schema.go:77]: Updated.

File: powerflex/sdc_expansion_resource.go [hunk 10]
```diff
@@ -0,0 +1,474 @@
...
+		return &parseCSVResponse, fmt.Errorf("Meesage : %s, Error Cosde : %s", parsecsvRespose.Message, strconv.Itoa(parsecsvRespose.StatusCode))
+	}
+
+	return parsecsvRespose, nil
+}
+
+// ValidateMDMOperation function for Vaidate the MDM credentials
+func ValidateMDMOperation(ctx context.Context, plan CsvAndMdmDataModel, gatewayClient *goscaleio.GatewayClient) (*goscaleio_types.GatewayResponse, error) {
+	mapData := map[string]interface{}{
+		"mdmUser":     "admin",
```
[the remark this answers is not in the document]
Comment (rounak-adhikary) [reply powerflex/sdc_expansion_resource.go:369]: Oh i see that this has been answered.

File: powerflex/sdc_expansion_resource.go
[the remark this answers is not in the document]
Comment (AnikaAgiwal2711) [reply powerflex/sdc_expansion_resource.go:1]: That's a good idea

File: powerflex/sdc_expansion_resource_schema.go [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+var SDCExpansionResourceSchema schema.Schema = schema.Schema{
+	Description:         "This resource can be used to add the SDC in PowerFlex Cluster.",
+	MarkdownDescription: "This resource can be used to add the SDC in PowerFlex Cluster.",
+	Attributes: map[string]schema.Attribute{
+		"csv_detail": csvSchema,
+		"mdm_ip": schema.StringAttribute{
```
[the remark this answers is not in the document]
Comment (shenda1) [reply powerflex/sdc_expansion_resource_schema.go:49]: @rounak-adhikary  Anyway, we are accepting the IPs with the role (primary, secondary) in the csv_detail parameter. Removing this attribute makes sense?

[the remark this answers is not in the document]
Comment (rounak-adhikary) [reply powerflex/sdc_expansion_resource_schema.go:49]: Ah yes, rt... then definitely we shd remove this.

File: powerflex/sdc_expansion_resource_schema.go [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+				Description:         "Password on the node",
+				Required:            true,
+				MarkdownDescription: "Password on the node",
+			},
+			"operating_system": schema.StringAttribute{
+				Description:         "Operating System on the node",
+				Required:            true,
+				MarkdownDescription: "Operating System on the node",
+			},
+			"is_mdm_or_tb": schema.StringAttribute{
```
Comment (rounak-adhikary) [inline powerflex/sdc_expansion_resource_schema.go:99]: This should have a validator for acceptable values, and the descriptions should reflect that.

File: powerflex/sdc_expansion_resource_schema.go
Comment (rounak-adhikary) [inline powerflex/sdc_expansion_resource_schema.go:1]: Please update the descriptions with default and acceptable values.

File: powerflex/sdc_expansion_resource_schema.go (hunk 7 above)
Comment (Krunal-Thakkar) [reply powerflex/sdc_expansion_resource_schema.go:48]: Updated.

File: powerflex/sdc_expansion_resource.go [hunk 13]
```diff
@@ -0,0 +1,474 @@
...
+			"unexpected error: "+queueOperationError.Error(),
+		)
+		return
+	}
+
+	resp.State.RemoveResource(ctx)
+}
+
+// QueueOpeartions function for the Abort, Clear and Move To Idle Execution
+func QueueOpeartions(gatewayClient *goscaleio.GatewayClient) error {
```
[the remark this answers is not in the document]
Comment (Krunal-Thakkar) [reply powerflex/sdc_expansion_resource.go:276]: Update.

File: powerflex/sdc_expansion_resource.go [hunk 14]
```diff
@@ -0,0 +1,474 @@
...
+						return fmt.Errorf("Error Clearing Queue During Installation is %s", queueOperationError.Error())
+					}
+
+					couterForStopExecution = 10
+
+					return nil
+				}
+
+			} else {
+				if checkForPhaseCompleted.Data == "Running" {
```
[the remark this answers is not in the document]
Comment (Krunal-Thakkar) [reply powerflex/sdc_expansion_resource.go:438]: Updated.

File: powerflex/sdc_expansion_resource.go [hunk 15]
```diff
@@ -0,0 +1,474 @@
...
+							return fmt.Errorf("Error Clearing Queue During Installation is %s", queueOperationError.Error())
+						}
+
+						return fmt.Errorf("Time Out,Some Operations of Installer running from since long")
+					}
+
+				} else {
+					if couterForStopExecution < 5 {
+
+						gatewayClient.RetryPhase()
```
[the remark this answers is not in the document]
Comment (Krunal-Thakkar) [reply powerflex/sdc_expansion_resource.go:454]: Updated.

File: powerflex/sdc_expansion_resource.go
[the remark this answers is not in the document]
Comment (Krunal-Thakkar) [reply powerflex/sdc_expansion_resource.go:1]: Yeah Sure.

File: powerflex/sdc_expansion_resource_schema.go [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+				Description:         "Whether this works as MDM or Tie Breaker",
+				Optional:            true,
+				MarkdownDescription: "Whether this works as MDM or Tie Breaker",
+				PlanModifiers: []planmodifier.String{
+					stringDefault(" "),
+				},
+			},
+			"is_sdc": schema.StringAttribute{
```
[the remark this answers is not in the document]
Comment (Krunal-Thakkar) [reply powerflex/sdc_expansion_resource_schema.go:124]: I follow with Gateway server properties.

File: powerflex/sdc_expansion_resource_schema.go
Comment (Krunal-Thakkar) [reply powerflex/sdc_expansion_resource_schema.go:1]: Updated.

File: powerflex/sdc_expansion_resource.go [hunk 17]
```diff
@@ -0,0 +1,474 @@
...
+	}
+
+	return validateMDMResponse, nil
+}
+
+// InstallationOperations function for begin instllation process
+func InstallationOperations(ctx context.Context, plan CsvAndMdmDataModel, gatewayClient *goscaleio.GatewayClient, parsecsvRespose *goscaleio_types.GatewayResponse) error {
+	beginInstallationResponse, _ := gatewayClient.BeginInstallation(parsecsvRespose.Data, "admin", plan.MdmPassword.ValueString(), plan.LiaPassword.ValueString(), true)
+
+	if beginInstallationResponse.StatusCode == 200 {
```
[the remark this answers is not in the document]
Comment (Krunal-Thakkar) [reply powerflex/sdc_expansion_resource.go:476]: I return with status code and error message from Gateway server, so based on status code If I need to put any logic, So I can put that.

**Review by shenda1 — DISMISSED** [review]

File: powerflex/sdc_expansion_resource_schema.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (shenda1) [reply powerflex/sdc_expansion_resource_schema.go:118]: By the attribute behavior, the boolean makes sense.

[the remark this answers is not in the document]
Comment (Krunal-Thakkar) [reply powerflex/sdc_expansion_resource_schema.go:127]: I have considered with Case Insensitive, so not make any problem with string.

[the remark this answers is not in the document]
Comment (rounak-adhikary) [reply powerflex/sdc_expansion_resource_schema.go:127]: Okay, we can have this for now.

**Review by rounak-adhikary — APPROVED** [review]
LGTM

**Pull request merged**